core/kotlinx-coroutines-test/src/TestCoroutineDispatcher.kt

```diff
@@ -0,0 +1,296 @@
+     */
+    @ExperimentalCoroutinesApi
+    fun pauseDispatcher()
+
+    /**
+     * Resume the dispatcher from a paused state.
+     *
+     * Resumed dispatchers will automatically progress through all coroutines scheduled at the current time. To advance
+     * time and execute coroutines scheduled in the future use one of [advanceTimeBy], [advanceTimeToNextDelayed],
+     * or [advanceUntilIdle].
+     */
+    @ExperimentalCoroutinesApi
+    fun resumeDispatcher()
+}
+
+/**
+ * Thrown when a test has completed by there are tasks that are not completed or cancelled.
+ */
+@ExperimentalCoroutinesApi
+class UncompletedCoroutinesError(message: String, cause: Throwable? = null): AssertionError(message, cause)
+
+/**
+ * [CoroutineDispatcher] that can be used in tests for both immediate and lazy execution of coroutines.
+ *
+ * By default, [TestCoroutineDispatcher] will be immediate. That means any tasks scheduled to be run without delay will
+ * be immediately executed. If they were scheduled with a delay, the virtual clock-time must be advanced via one of the
+ * methods on [DelayController]
+ *
+ * When swiched to lazy execution any coroutines started via [launch] or [async] will
+ * not execute until a call to [DelayController.runCurrent] or the virtual clock-time has been advanced via one of the
+ * methods on [DelayController].
+ *
+ * @see DelayController
+ */
+@ExperimentalCoroutinesApi
+class TestCoroutineDispatcher:
+        CoroutineDispatcher(),
+        Delay,
+        DelayController {
+
+    private var dispatchImmediately = true
+        set(value) {
+            field = value
+            if (value) {
+                // there may already be tasks from setup code we need to run
+                advanceUntilIdle()
+            }
+        }
+
+    // The ordered queue for the runnable tasks.
+    private val queue = ThreadSafeHeap<TimedRunnable>()
+
+    // The per-scheduler global order counter.
+    private var counter = 0L
+
+    // Storing time in nanoseconds internally.
+    private var time = 0L
+
+    override fun dispatch(context: CoroutineContext, block: Runnable) {
+        if (dispatchImmediately) {
+            block.run()
+        } else {
+            post(block)
+        }
+    }
+
+    override fun scheduleResumeAfterDelay(timeMillis: Long, continuation: CancellableContinuation<Unit>) {
+        postDelayed(CancellableContinuationRunnable(continuation) { resumeUndispatched(Unit) }, timeMillis)
+    }
+
+    override fun invokeOnTimeout(timeMillis: Long, block: Runnable): DisposableHandle {
+        val node = postDelayed(block, timeMillis)
+        return object : DisposableHandle {
+            override fun dispose() {
+                queue.remove(node)
+            }
+        }
+    }
+
+    override fun toString(): String = "TestCoroutineDispatcher[time=${time}ns]"
+
+    private fun post(block: Runnable) =
+            queue.addLast(TimedRunnable(block, counter++))
```

Comment:
looks like `post` might be called in any thread (via dispatch). should the `counter` be atomic ? alternatively, all posts could be enqueued into a separate thread?

Reply:
Migrated it to use Atomic, which will make the bulk of the problems go away. Need to make sure block.run in dispatch doesn't accidentally run things on weird threads too.